python/hail/methods/statgen.py

```diff
@@ -204,6 +204,729 @@ def linreg(dataset, ys, x, covariates=[], root='linreg', block_size=16):
+    Firth introduced his approach in
+    `Bias reduction of maximum likelihood estimates, 1993 <http://www2.stat.duke.edu/~scs/Courses/Stat376/Papers/GibbsFieldEst/BiasReductionMLE.pdf>`__.
+    Heinze and Schemper further analyze Firth's approach in
+    `A solution to the problem of separation in logistic regression, 2002 <https://cemsiis.meduniwien.ac.at/fileadmin/msi_akim/CeMSIIS/KB/volltexte/Heinze_Schemper_2002_Statistics_in_Medicine.pdf>`__.
+
+    Those variants that don't vary across the included samples (e.g., all
+    genotypes are HomRef) will have missing annotations.
+
+    Phenotype and covariate sample annotations may also be specified using
+    `programmatic expressions <exprlang.html>`__ without identifiers, such as:
+
+    .. code-block:: text
+
+        if (ds.isFemale) ds.cov.age else (2 * ds.cov.age + 10)
+
+    For Boolean covariate types, true is coded as 1 and false as 0. In
+    particular, for the sample annotation `fam.isCase` added by importing a FAM
+    file with case-control phenotype, case is 1 and control is 0.
+
+    Hail's logistic regression tests correspond to the ``b.wald``, ``b.lrt``,
+    and ``b.score`` tests in `EPACTS`_. For each variant, Hail imputes missing
+    input values as the mean of non-missing input values, whereas EPACTS
+    subsets to those samples with called genotypes. Hence, Hail and EPACTS
+    results will currently only agree for variants with no missing genotypes.
+
+    .. _EPACTS: http://genome.sph.umich.edu/wiki/EPACTS#Single_Variant_Tests
+
+    Parameters
+    ----------
+    test : {'wald', 'lrt', 'score', 'firth'}
+        Statistical test.
+    y : :class:`.Expression`
+        Response expression. Must evaluate to Boolean or numeric with all values
+        0 or 1.
+    x : :class:`.Expression`
+        Expression for input variable.
+    covariates : :obj:`list` of :class:`.Expression`, optional
+        Covariate expressions.
+    root : :obj:`str`, optional
+        Name of resulting row-indexed field.
+
+    Returns
+    -------
+    :class:`.MatrixTable`
+        Dataset with regression results in a new row-indexed field.
+    """
+
+    # x is entry-indexed
+    analyze('logreg/x', x, dataset._entry_indices)
+
+    # y and covariates are col-indexed
+    analyze('logreg/y', y, dataset._col_indices)
+
+    all_exprs = [x, y]
+    for e in covariates:
+        all_exprs.append(e)
+        analyze('logreg/covariates', e, dataset._col_indices)
+
+    base, cleanup = dataset._process_joins(*all_exprs)
+
+    jds = base._jvds.logreg(
+        test,
+        y._ast.to_hql(),
+        x._ast.to_hql(),
+        jarray(Env.jvm().java.lang.String,
+               [cov._ast.to_hql() for cov in covariates]),
+        'va.`{}`'.format(root))
+
+    return cleanup(MatrixTable(jds))
+
+@handle_py4j
+@typecheck(ds=MatrixTable,
+           kinshipMatrix=KinshipMatrix,
+           y=Expression,
```

Comment:
numeric_expr? (I can't tell from the parameters list, but I assume they have to be numeric-ish? If so, probably should also make that clear in the parameters list.)